### ISSUE TITLE
Does opensourcepos has credits payments and outstanding balance support?

### ISSUE THREAD

**Issue by abuzer** [jekkos/opensourcepos#1428]
Hi, 
I am wandering if opensourcepos has credits payments to track sales on credits and a report to list all out standing payment from a customer.

Thanks

**Comment by guillermohaad** [comment]
Hi, @abuzer 
Does not handle credits or payments, same question on ticket # 1427

**Comment by daN4cat** [comment]
Needs to be added, please contribute it as you cannot expect the 2 or 3 of us to implement everything....

**Comment by guillermohaad** [comment]
ok @daN4cat Will inform

**Comment by asadjaved63** [comment]
@daN4cat  i want to contribute debit credits payments system , can you tell me some logic  how to work system?

**Comment by daN4cat** [comment]
@asadjaved63 it's not a one line answer. First you need to get familiar with the code and read the wiki for database relationship table design and get familiar with CodeIgniter MVC concept.
You would need to check the code in Controllers and Models, things like Sales and Sale.
That's as starting point.
Maybe @SteveIreland can give some more hints because he recently started contributing and therefore has a refresher view of this. I look into OPSOP and the predecessor for a number of years so I know where to go to change and fix something... outside the js area that is still a bit far from me.

**Comment by SteveIreland** [comment]
A great man by the name of Sergeant  Schultz once said "I know nothing - I see nothing - I hear nothing!".

Actually he said it in just about every episode of Hogan's Heroes.  I lived by those words since I was 15 years old.

**Comment by SteveIreland** [comment]
Seriously though, I raised this same request (as have many others before me) and came to the same conclusion as others before me that a back office app is the better solution because one you create an an invoice then you've place a legally binding debt on your customer that they are obligated to pay lest they get in to legal trouble, but that also means that you need to track that debt and send a monthly statement to them reporting the outstanding balance of their debt. 

That's why I'm starting to work on integrating FrontAccounting with OSPOS.  FrontAccounting tracks debt and generates statements.

**Comment by daN4cat** [comment]
Yep, Steve, you have a good point there. Totally agree.
However I'm puzzled by the invoice life cycle in OSPOS then and what a payment means.
Having said that, I don't use invoices or the rare cases we do we just manage outside OSPOS removing the item quantities from the stock.

**Comment by daN4cat** [comment]
But going back to the topic here, it's how to manage credit payment without invoices and as a small shop.

**Comment by SteveIreland** [comment]
In our scenario I've adapted quotes to behave like work orders and the work order is used to track services and parts (I still have opportunities for improvement which I am continuing to work on).  Then when the work is completed we "unsuspend" the work order and convert it to an invoice.  Deposit payments have been made (either via cash or credit) which were entered before the actual work started - I added to the payment types both "Cash deposit" and "Credit deposit".  At that point I expect the invoice to be sent to FrontAccounting, along with the deposit payments.  Then from that point FrontAccounting should be used to generate statements for the customer of the amount due and used to post additional payments.

I haven't decided yet if I need to update OSPOS to reflect those post transaction payments.

**Comment by amrshah** [comment]
Hi,

I am trying to achieve sale on "credit" and keep a track of "outstanding balance" as well for the customer.

I have this approach to discuss if it will be right to do so:

1. Define a sales_payment_type = "On Account"
2a. Have a column named "Balance" on customer table with a track of the outstanding amount
2b. Have a "Balance" attribute on customer and calculate it based on payment type "On Account" and report back on the customer view? Just a calculation like the one already implemented where we are getting the customer Totals amount that tells the total transactions volume with this customer. (Ref: get_totals($customer_id) method) we could add a get_balance($customer_id) method to do so.
2c. Add an Account table keeping track of the outstanding balance.

For simple tracking of the account balance we could do one of the above. However a full/proper integration of payment module or integration with frontaccounting is the best methodology to follow.

Any insights?

**Comment by amrshah** [comment]
One simplest Option could be to have a get_totals($customer_id, $payment_type) method.

we could query with $payment_type = "On Account" // or "Credit Card", "Debit Card" etc

This can provide an extra level of detail on Customers Payment history/totals and solve the balance tracking feature/requirement as well :)

**Comment by daN4cat** [comment]
@amrshah There is a payment "Due" that is the Credit payment and it's in latest master. Let's build on top of that.

**Comment by daN4cat** [comment]
Also Takings has a filter for Due payments so you can see what's left to be paid.
However you may want to add a specific report to get due payments grouped by customer names with a detailed list of each singular transaction (like in detailed sales).

**Comment by amrshah** [comment]
Yes thank you @daN4cat I have read about that just today. Alright it is good to build on top of that. I am brainstorming on this and related things like settlement of the Due amount at a later time, how should that be done and so on.

Remember the Sales report problem I posted yesterday. I am trying to figure out how to fix these too on my XAMP 7.
I don't have my machine setup with all the tools (docker, eclipse etc). I m using Netbeans IDE for PHP for now.  But I am interested to work on this thing :)

Thanks again for your suport

**Comment by amrshah** [comment]
![takings](https://user-images.githubusercontent.com/1771447/29486520-042b3dba-8501-11e7-9d8c-35e956183e71.png)

See the takings screen? The drop down isn't listing other payment types

**Comment by daN4cat** [comment]
I'm sorry for you but I'm afraid latest master has what I said it has because I implemented it:

![image](https://user-images.githubusercontent.com/636048/29486657-48a80db0-84e3-11e7-8ea1-e3b10e6ed906.png)

**Comment by daN4cat** [comment]
And that's 3.1.0 not 3.0.2 that you are using. I would discourage implementing new features on old releases as 3.1.0 is close to be released.

**Comment by amrshah** [comment]
Yes agreed. I am just trying to set it up so I can do some real work.  I couldn't get the latest one working so I had installed 3.0.2. I am trying to set up the latest one in the meanwhile. 
Sorry for inconvenience

**Comment by amrshah** [comment]
What is the Code Igniter version with master? I downgraded the system folder to earlier CI versions which now gives me a blank page. I have gone through the reported issues of Server Error 500 and rest of the WIKI, FAQ etc. Everything looks fine. May be there's a problem with CI version? Shouldn't be but just trying to find out...tired now :-S

**Comment by amrshah** [comment]
Here is the Apache log:

```
[Sun Aug 20 00:10:12.506987 2017] [mpm_winnt:notice] [pid 7468:tid 444] AH00418: Parent: Created child process 6584
[Sun Aug 20 00:10:13.027874 2017] [ssl:warn] [pid 6584:tid 464] AH01909: www.example.com:443:0 server certificate does NOT include an ID which matches the server name
[Sun Aug 20 00:10:13.134654 2017] [ssl:warn] [pid 6584:tid 464] AH01909: www.example.com:443:0 server certificate does NOT include an ID which matches the server name
[Sun Aug 20 00:10:13.159220 2017] [mpm_winnt:notice] [pid 6584:tid 464] AH00354: Child: Starting 150 worker threads.
```

**Comment by amrshah** [comment]
Something related to SSL?

**Comment by amrshah** [comment]
After changing the listen port 80 and listen port 443 in apache configs, the errorstill not resolved :-S

**Comment by daN4cat** [comment]
Not sure why you fiddle with code igniter versions. That has nothing to do with your issues.

**Comment by daN4cat** [comment]
The SSL issues you get are due to a problem with your certificate and probably the fact that you set www.example.com, which is a protected domain.

**Comment by amrshah** [comment]
`Not sure why you fiddle with code igniter versions. That has nothing to do with your issues.
`
I remember facing same error a couple years ago while I was working with codeigniter. Secondly the error occurs when I place codeigniter's system folder under vendor/framework. So that's why I was thinking there might be something to do with CI.

`What is the Code Igniter version with master? I downgraded the system folder to earlier CI versions which now gives me a blank page.`

?

**Comment by daN4cat** [comment]
I don't get why you are doing that. It's not supposed to be like that.

If you are downloading the code from Github you need to follow the developer installation. Read the wiki, as you need bower, grunt and composer. Then run the devinstall script under bin to get all installed.
If not from github take the zip file from bintray as indicated in the Readme (unstable versions).

You don't need to wonder about code igniter it's all sorted.

**Comment by guillermohaad** [comment]
friends where can I translate the type of payment sales_due? I have set up in Spanish and it continues to come out in English.

![image](https://user-images.githubusercontent.com/12921805/31046941-6dcd7dfa-a5c7-11e7-9aef-acca1f523147.png)


I thank a charitable soul

**Comment by daN4cat** [comment]
@guillermohaad it's missing the Spanish translation for those strings, the "english" is not actually a translation but the name of the string you need to translate

**Comment by mckaygerhard** [comment]
only related but not similar/specific work: see #2080 (will be realted to xpenses)
